src/Integration/Binding/NuGetBindingOperation.cs

```diff
@@ -0,0 +1,148 @@
+
+            if (projectsToBind == null || !projectsToBind.Any())
+            {
+                Debug.Fail("Not expected to be called when there are no projects");
+                return true;
+            }
+
+            var projectNugets = projectsToBind
+                .SelectMany(bindingProject =>
+                {
+                    var projectLanguage = Language.ForProject(bindingProject);
+
+                    List<NuGetPackageInfoResponse> nugetPackages;
+                    if (!this.NuGetPackages.TryGetValue(projectLanguage, out nugetPackages))
+                    {
+                        var message = string.Format(Strings.BindingProjectLanguageNotMatchingAnyQualityProfileLanguage, bindingProject.Name);
+                        this.logger.WriteLine(Strings.SubTextPaddingFormat, message);
+                        nugetPackages = new List<NuGetPackageInfoResponse>();
+                    }
+
+                    return nugetPackages.Select(nugetPackage => new { Project = bindingProject, NugetPackage = nugetPackage });
+                })
+                .ToArray();
+
+            bool overallSuccess = true;
+
+            DeterminateStepProgressNotifier progressNotifier = new DeterminateStepProgressNotifier(notificationEvents, projectNugets.Length);
+            foreach (var projectNuget in projectNugets)
+            {
+                if (token.IsCancellationRequested)
+                {
+                    break;
+                }
+
+                string message = string.Format(CultureInfo.CurrentCulture, Strings.EnsuringNugetPackagesProgressMessage, projectNuget.NugetPackage.Id, projectNuget.Project.Name);
+                this.logger.WriteLine(Strings.SubTextPaddingFormat, message);
+
+                var isNugetInstallSuccessful = NuGetHelper.TryInstallPackage(this.serviceProvider, this.logger, projectNuget.Project, projectNuget.NugetPackage.Id, projectNuget.NugetPackage.Version);
+
+                if (isNugetInstallSuccessful) // NuGetHelper.TryInstallPackage already displayed the error message so only take care of the success message
+                {
+                    message = string.Format(CultureInfo.CurrentCulture, Strings.SuccessfullyInstalledNugetPackageForProject, projectNuget.NugetPackage.Id, projectNuget.Project.Name);
+                    this.logger.WriteLine(Strings.SubTextPaddingFormat, message);
+                }
+
+                // TODO: SVS-33 (https://jira.sonarsource.com/browse/SVS-33) Trigger a Team Explorer warning notification to investigate the partial binding in the output window.
+                overallSuccess &= isNugetInstallSuccessful;
+
+                progressNotifier.NotifyIncrementedProgress(string.Empty);
+            }
+            return overallSuccess;
+        }
+
+        public bool ProcessExport(Language language, RoslynExportProfileResponse exportProfileResponse)
```

Comment:
The code in this method was extracted from inside _BindingWorkflow.DownloadQualityProfileAsync_.